### PR TITLE
[JENKINS-40872] Respond with HTTP 307 before redirect to login page

### PR DESCRIPTION
**Pull request by jdussouillez**
See [JENKINS-40872](https://issues.jenkins-ci.org/browse/JENKINS-40872).
Respond with HTTP "Temporary redirect" instead of "Forbidden" before redirect to login page.
Respond with HTTP "Forbidden" only if it's a AJAX request (no redirection).

### Proposed changelog entries

* `Internal:` Respond with HTTP "Temporary redirect" before redirect to login page

### Submitter checklist

- [x] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@rtyler (JENKINS-40872 author)

### CODE REVIEW

**Comment by daniel-beck** [comment]
JENKINS-40872 is specifically about the root URL, while this changes any URL. I am concerned that this will break programmatic use cases. This will require extensive testing.

**Review by oleg-nenashev — COMMENTED** [review]
No strong opinion here. I agree with comments from @daniel-beck , careful testing is needed. CC @jenkinsci/code-reviewers @jenkinsci/core

File: core/src/main/java/hudson/security/HudsonAuthenticationEntryPoint.java [hunk 1]
```diff
@@ -80,7 +81,7 @@ public void commence(ServletRequest request, ServletResponse response, Authentic
             loginForm = MessageFormat.format(loginForm, URLEncoder.encode(uriFrom,"UTF-8"));
             req.setAttribute("loginForm", loginForm);
 
-            rsp.setStatus(SC_FORBIDDEN);
+            rsp.setStatus(HttpServletResponse.SC_TEMPORARY_REDIRECT);
```
Comment (jtnord) [inline core/src/main/java/hudson/security/HudsonAuthenticationEntryPoint.java]: regardless of any other comments - where does the `Location` header get set that the client should use for the redirect?  I could not see it in the code - and if performing a redirect like this then you shouldn't (need to) include the HTML that does the redirect below.

Comment (jtnord) [reply core/src/main/java/hudson/security/HudsonAuthenticationEntryPoint.java]: Also I think this should be using `302` - as we do not want the browser to POST if this was from a POST request at the new URL...

**Comment by batmat** [comment]
Also, 4 test failures that definitely look related.
image

**Comment by daniel-beck** [comment]
It doesn't look like this is going to make progress any time soon, so closing for now.

Please feel free to resubmit this change if you're still interested in it.

**Pull request closed without merging**